graylog2-server/src/main/java/org/graylog2/security/InMemoryRolePermissionResolver.java

```diff
@@ -0,0 +1,106 @@
+/**
+ * This file is part of Graylog.
+ *
+ * Graylog is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Graylog is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.security;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Collections2;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import org.apache.shiro.authz.Permission;
+import org.apache.shiro.authz.permission.RolePermissionResolver;
+import org.apache.shiro.authz.permission.WildcardPermission;
+import org.graylog2.shared.users.Role;
+import org.graylog2.users.RoleService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+@Singleton
+public class InMemoryRolePermissionResolver implements RolePermissionResolver {
+    private static final Logger log = LoggerFactory.getLogger(InMemoryRolePermissionResolver.class);
+
+    private final RoleService roleService;
+    private final AtomicReference<ImmutableMap<String, Role>> idToRoleIndex = new AtomicReference<>(ImmutableMap.<String, Role>of());
+
+    @Inject
+    public InMemoryRolePermissionResolver(RoleService roleService,
+                                          @Named("daemonScheduler") ScheduledExecutorService daemonScheduler) {
+        this.roleService = roleService;
+        final RoleUpdater updater = new RoleUpdater();
+
+        // eagerly load rules
+        updater.run();
+
+        // update rules every second in the background
+        daemonScheduler.scheduleAtFixedRate(updater, 1, 1, TimeUnit.SECONDS);
```

Comment:
This is such hack. :disappointed:
(But yeah, we have to live with it right now)